crates/bevy_ecs/src/schedule/fsm.rs

```diff
@@ -0,0 +1,80 @@
+use crate::{
+    self as bevy_ecs,
+    change_detection::Mut,
+    schedule::{run_scheduled, SystemLabel},
+    system::Res,
+    world::World,
+};
+
+use std::{fmt::Debug, hash::Hash};
+use thiserror::Error;
+
+/// Types that can define states in a finite-state machine.
+pub trait State: 'static + Send + Sync + Clone + PartialEq + Eq + Debug + Hash {}
+impl<T> State for T where T: 'static + Send + Sync + Clone + PartialEq + Eq + Debug + Hash {}
+
+/// A [`SystemLabel`] for the system set that runs during a state's "on enter" transition.
+#[derive(Debug, Clone, Hash, PartialEq, Eq, SystemLabel)]
+pub struct OnEnter<S: State>(pub S);
+
+/// A [`SystemLabel`] for the system set that runs during a state's "on exit" transition.
+#[derive(Debug, Clone, Hash, PartialEq, Eq, SystemLabel)]
+pub struct OnExit<S: State>(pub S);
```

Comment:
I wonder how we make this work with #4957...

Reply:
What if we turn the `State` trait into another variant of the `define_label` macro? There's precedent for states not being able to store extra data, since it used to use `std::mem::discriminant` for comparisons. Since we're redesigning the API anyway it would be a good time to introduce a breaking change.

It makes sense -- if a `State` can act like a `SystemLabel`, it should have to follow the rules of one too. #5377 would also help out with this.

Reply:
or an even better solution: just require users to `#[derive(State)]`. Since we already require six other derives in order for `State` to be auto-`impl`'d, it would be more clear to just yeet those and do a custom derive for `State`.